blog/2024-04-03-monadic-notation-for-rust-translation.md

```diff
@@ -0,0 +1,166 @@
+
+The notation&nbsp;`f (| x1, ..., xn |)` represents the call to the function&nbsp;`f` with the arguments&nbsp;`x1`, ..., `xn` returning a monadic result. We bind the result with the current continuation that goes up to the wrapping `ltac:(M.monadic ...)` tactic. We automatically transform the `let` into a `let*` with the `M.monadic` tactic when needed.
+
+## Where do we use this notation? 🤔
+
+We use this notation in all the function bodies that we generate, that are all in a monad to represent side effects. We call the `ltac:(M.monadic ...)` tactic at the start of the functions, as well as at the start of closure bodies that are defined inside functions. This also applies to the translation of `if`, `match`, and `loop` expressions, as we represent their bodies as functions.
+
+Here is an example of code with a `match` expression:
+
+```rust
+fn add(a: i32, b: i32) -> i32 {
+    match a - b {
+        0 => a + b,
+        _ => a - b,
+    }
+}
+```
+
+We translate it to:
+
+```coq
+Definition add (τ : list Ty.t) (α : list Value.t) : M :=
+  match τ, α with
+  | [], [ a; b ] =>
+    ltac:(M.monadic
+      (let a := M.alloc (| a |) in
+      let b := M.alloc (| b |) in
+      M.read (|
+        M.match_operator (|
+          M.alloc (| BinOp.Panic.sub (| M.read (| a |), M.read (| b |) |) |),
+          [
+            fun γ =>
+              ltac:(M.monadic
+                (let _ :=
+                  M.is_constant_or_break_match (|
+                    M.read (| γ |),
+                    Value.Integer Integer.I32 0
+                  |) in
+                M.alloc (|
+                  BinOp.Panic.add (| M.read (| a |), M.read (| b |) |)
+                |)));
+            fun γ =>
+              ltac:(M.monadic (
+                M.alloc (|
+                  BinOp.Panic.sub (| M.read (| a |), M.read (| b |) |)
+                |)
+              ))
+          ]
+        |)
+      |)))
+  | _, _ => M.impossible
+  end.
+```
+
+We see that we call the tactic `M.monadic` for each branch of the `match` expression.
+
+## How does it work? 🛠️
+
+The `M.monadic` tactic is defined in [M.v](https://github.com/formal-land/coq-of-rust/blob/main/CoqOfRust/M.v). The main part is:
+
+```coq showLineNumbers
+Ltac monadic e :=
+  lazymatch e with
+  (* ... *)
+  | context ctxt [M.run ?x] =>
+    lazymatch context ctxt [M.run x] with
+    | M.run x => monadic x
+    | _ =>
+      refine (M.bind _ _);
+        [ monadic x
+        | let v := fresh "v" in
+          intro v;
+          let y := context ctxt [v] in
+          monadic y
+        ]
+    end
+  (* ... *)
+  end.
+```
+
+The `M.run` function is an axiom that we use as a marker to know where we need to apply the monadic bind of type `M A -> (A -> M B) -> M B`. The type of `M.run` is:
```

Comment:
Note that with the current handling of Rust types in actually has the type `M -> Value.t`

Reply:
And for the bind operation it is `M -> (Value.t -> M) -> M`